rfcs/text/0000-dss-bundle-enumeration.md

```diff
@@ -0,0 +1,44 @@
+### DCP PR:
+
+***Leave this blank until the RFC is approved** then the **Author(s)** must create a link between the assigned RFC number and this pull request in the format:*
+
+`[dcp-community/rfc#](https://github.com/HumanCellAtlas/dcp-community/pull/<PR#>)`
+
+# RFC: Bundle Enumeration of the DCP Data Store (DSS)
+
+## Summary
+
+This RFC proposes bundle enumeration endpoint(s) for the DSS.
+
+## Author(s)
+
+* [Brian Hannafious](mailto:bhannafi@ucsc.edu)
+
+## Shepherd
+***Leave this blank.** This role is assigned by DCP PM to guide the **Author(s)** through the RFC process.*
+
+*Recommended format for Shepherds:*
+
+ `[Name](mailto:username@example.com)`
+
+## Motivation
+
+Currently, the only means of listing bundles stored in the DSS is the internal Elasticsearch (ES) metadata index, which
+must be kept current with object storage. The DSS should provide bundle enumeration independently of the ES metadata index,
+emphasizing consistency and scalability.
+
+### User Stories
+
+* As a downstream service developer, I would like to enumerate the bundle contents of the DSS so I can create my own
+  index.
+
+* As a downstream service developer, I would like to check if my index contains all the bundles in the DSS.
+
+## Detailed Design
+
+A new bundle enumeration endpoint, `GET /bundles`, will be introduced, taking replica and prefix parameters. A listing
+all all bundles having UUIDs beginning with `prefix` will be returned directly from object storage. The prefix
+parameter is intended to facilitate parallelized listings. Pagination semantics and all other semantics of this route
+will be in line with the established conventions of the DSS API.
+
```

Comment:
Without some kind of filtering, this seems like a very heavyweight endpoint to use.  A couple of things come to mind:
- Filter by update date/time.  This would allow only obtaining bundles that have changes since the last time the query was run.
- Filter by bundle type.  Well, first we have to have bundle types.

Reply:
This endpoint is intended for heavyweight use by downstream indexers.

Also, an incremental approach seems preferable: if filtering becomes desirable in the future, it can be added to the endpoint.

Reply:
Agreed that this is easy to add downstream.  Assuming a full dump is what existing users want, then my speculative use is not a real use case ;-)

Reply:
@diekhans you raise a good point, but as @xbrianh pointed out, the use case here is an unfiltered bulk pull of all bundle IDs for external indexing. We did look for a way to use a "lightweight" database to do filtering using our established filter language process (JMESpath), but didn't find any suitable database/indexing engine for such a task.